### PR TITLE
google-cloud-core: Optimize Date.parseDate(String)

### PR DESCRIPTION
**Pull request by olavloite**
Date parsing using a regular expression is slower than a specific implementation for the only format that is supported.

`com.google.cloud.Date` supports invalid dates. A date like `2001-02-29` is accepted, both when you submit it through `Date.parseDate` as well as by constructing it using the `Date.fromYearMonthDay`. This behavior is not changed by this PR.

### CODE REVIEW

File: google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/IntParser.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  static int parseInt(String input, int begin, int end, int exponent) {
+    Preconditions.checkNotNull(input);
+    Preconditions.checkArgument(
+        exponent >= 0 && exponent < POWERS_OF_10.length, "Exponent out of range");
+    Preconditions.checkArgument(end - begin <= 10, "Max input length is 10");
+    Preconditions.checkArgument(end >= begin, "End must be greater or equal to begin");
+    Preconditions.checkArgument(begin >= 0, "Begin must be >= 0");
+    Preconditions.checkArgument(end <= input.length(), "End must be <= input.length()");
+    int res = 0;
+    for (int index = begin; index < end; index++) {
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/IntParser.java]: Is there a reason not to call `Integer.parseInt(input.substring(begin, end)` here?

File: google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java [hunk 2]
```diff
@@ -34,10 +35,44 @@
...
+    parseInvalidDate("2016/09/18");
+    parseInvalidDate("2016 09 18");
+    parseInvalidDate("2016-9-18");
+    parseInvalidDate("2016-09-18T10:00");
+    parseInvalidDate("");
+    parseInvalidDate("test");
+    parseInvalidDate("2000-13-01");
+    parseInvalidDate("2000-12-32");
+    parseInvalidDate("10000-01-01");
+    parseInvalidDate("0000-01-01");
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java]: please add `aaaa-bb-cc`

File: google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/IntParser.java (hunk 1 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-core/src/main/java/com/google/cloud/IntParser.java]: No, good point. Changed.

File: google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java (hunk 2 above)
Comment (olavloite) [reply google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java]: Added.

File: google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java [hunk 3]
```diff
@@ -34,10 +35,51 @@
...
+    parseInvalidDate("0001--1-01");
+    parseInvalidDate("0001-01--1");
+  }
+
+  private void parseInvalidDate(String input) {
+    try {
+      Date.parseDate(input);
+      fail("Expected exception");
+    } catch (IllegalArgumentException e) {
+      // Ignore, this is the expected exception.
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java]: The message should contain `"Invalid date"`

Comment (olavloite) [reply google-cloud-clients/google-cloud-core/src/test/java/com/google/cloud/DateTest.java]: Added try-catch for `NumberFormatException` and check for `"Invalid date"` in the exception message.

**Pull request merged**